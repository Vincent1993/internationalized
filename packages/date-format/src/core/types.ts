import type { ReactNode } from 'react';
import type { DateInput } from '../utils/types';

export interface DateFormatContextValue extends Intl.DateTimeFormatOptions {
  locale?: string | string[];
}

export interface DateFormatProviderProps {
  children: ReactNode;
  options?: DateFormatContextValue;
}

export interface UseDateFormatOptions extends Intl.DateTimeFormatOptions {
  locale?: string | string[];
}

export interface CreateDateFormatOptions extends UseDateFormatOptions {
  contextDefaults?: DateFormatContextValue | null;
}

export interface ResolvedDateFormatOptions extends Intl.ResolvedDateTimeFormatOptions {
  originalLocale?: string | string[];
}

export interface UseDateFormatResult {
  formattedValue: string;
  parts: Intl.DateTimeFormatPart[];
  date: Date | null;
  resolvedOptions: ResolvedDateFormatOptions;
  isInvalid: boolean;
}

export interface DateFormatter {
  format(value: DateInput, overrideOptions?: UseDateFormatOptions): UseDateFormatResult;
  resolveOptions(overrideOptions?: UseDateFormatOptions): ResolvedDateFormatOptions;
}

export type { DateInput };
