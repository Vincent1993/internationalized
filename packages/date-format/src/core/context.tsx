import { createContext, useContext, useMemo } from 'react';
import type { PropsWithChildren } from 'react';
import type { DateFormatContextValue, DateFormatProviderProps } from './types';

const DateFormatContext = createContext<DateFormatContextValue | null>(null);

/**
 * @name DateFormatProvider
 * @description 提供日期时间格式化上下文的 React 组件。它会缓存传入的 `options`
 * 并向所有子组件、Hook 共享，便于在应用内统一 locale、时区等配置。
 *
 * @since 0.0.1
 *
 * @example
 * ```tsx
 * import { DateFormatProvider, DateFormat } from '@internationalized/date-format';
 *
 * export function App() {
 *   return (
 *     <DateFormatProvider options={{ locale: 'zh-CN', timeZone: 'Asia/Shanghai' }}>
 *       <DateFormat value={new Date()} timeStyle="short" dateStyle="medium" />
 *     </DateFormatProvider>
 *   );
 * }
 * ```
 */
export function DateFormatProvider({ children, options }: DateFormatProviderProps) {
  const memoized = useMemo(() => options ?? null, [options]);

  return <DateFormatContext.Provider value={memoized}>{children}</DateFormatContext.Provider>;
}

/**
 * @name useDateFormatContext
 * @description 读取最近的 `DateFormatProvider` 上下文，供自定义组件或 Hook 访问日期配置。
 *
 * @since 0.0.1
 *
 * @returns 当前上下文配置；若不存在 Provider，则返回 `null`
 */
export function useDateFormatContext(): DateFormatContextValue | null {
  return useContext(DateFormatContext);
}

export type { DateFormatContextValue };
export type DateFormatProviderComponentProps = PropsWithChildren<DateFormatProviderProps>;
