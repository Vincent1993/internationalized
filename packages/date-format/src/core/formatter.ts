import { normalizeDateInput, normalizeLocale } from '../utils/date';
import type {
  CreateDateFormatOptions,
  DateFormatter,
  DateFormatContextValue,
  ResolvedDateFormatOptions,
  UseDateFormatOptions,
  UseDateFormatResult,
} from './types';
import type { DateInput } from '../utils/types';

interface MergeResult {
  locale?: string | string[];
  intlOptions: Intl.DateTimeFormatOptions;
  originalLocale?: string | string[];
}

function ensureLocale(locale?: string | string[] | null): string | string[] {
  const normalized = normalizeLocale(locale ?? undefined);
  return normalized ?? 'zh-CN';
}

type PartialOptionSource = UseDateFormatOptions | DateFormatContextValue | null | undefined;

function splitOptions(source: PartialOptionSource): MergeResult {
  if (!source) {
    return { intlOptions: {} };
  }

  const { locale, ...intlOptions } = source;

  return {
    locale: locale ? normalizeLocale(locale) : undefined,
    intlOptions: { ...intlOptions },
    originalLocale: locale ? normalizeLocale(locale) : undefined,
  };
}

function mergeOptions(
  contextOptions?: PartialOptionSource,
  baseOptions?: PartialOptionSource,
  overrideOptions?: PartialOptionSource,
): MergeResult {
  const merged: MergeResult = { intlOptions: {} };
  const sources: MergeResult[] = [
    splitOptions(contextOptions),
    splitOptions(baseOptions),
    splitOptions(overrideOptions),
  ];

  for (const source of sources) {
    if (source.locale !== undefined) {
      merged.locale = source.locale;
      merged.originalLocale = source.originalLocale;
    }

    Object.assign(merged.intlOptions, source.intlOptions);
  }

  return merged;
}

function buildResolvedOptions(
  locale: string | string[] | undefined,
  intlOptions: Intl.DateTimeFormatOptions,
  originalLocale?: string | string[],
): ResolvedDateFormatOptions {
  const normalizedLocale = ensureLocale(locale ?? originalLocale);
  const formatter = new Intl.DateTimeFormat(normalizedLocale, intlOptions);
  const resolved = formatter.resolvedOptions();

  const fallbackOriginal = originalLocale ?? normalizedLocale;

  return {
    ...resolved,
    originalLocale: fallbackOriginal,
  };
}

function formatValue(
  value: DateInput,
  locale: string | string[] | undefined,
  intlOptions: Intl.DateTimeFormatOptions,
  originalLocale?: string | string[],
): UseDateFormatResult {
  const normalizedDate = normalizeDateInput(value);
  const resolvedOptions = buildResolvedOptions(locale, intlOptions, originalLocale);

  if (!normalizedDate) {
    return {
      formattedValue: 'Invalid Date',
      parts: [],
      date: null,
      resolvedOptions,
      isInvalid: true,
    };
  }

  const normalizedLocale = ensureLocale(locale ?? originalLocale);
  const formatter = new Intl.DateTimeFormat(normalizedLocale, intlOptions);
  const parts = formatter.formatToParts(normalizedDate);

  return {
    formattedValue: formatter.format(normalizedDate),
    parts,
    date: normalizedDate,
    resolvedOptions,
    isInvalid: false,
  };
}

/**
 * @name resolveDateFormatOptions
 * @description 合并调用参数与上下文默认值，返回 `Intl.DateTimeFormat` 的完整解析结果。
 *
 * @since 0.0.1
 *
 * @param options - 本次调用的格式化配置
 * @param contextDefaults - 来自 `DateFormatProvider` 的上下文默认值
 * @returns `Intl.DateTimeFormat` 的解析选项，额外包含 `originalLocale`
 *
 * @example
 * ```ts
 * import { resolveDateFormatOptions } from '@internationalized/date-format';
 *
 * const resolved = resolveDateFormatOptions(
 *   { dateStyle: 'long', timeZone: 'Asia/Shanghai' },
 *   { locale: 'zh-CN' }
 * );
 *
 * console.log(resolved.locale); // zh-CN
 * console.log(resolved.timeZone); // Asia/Shanghai
 * console.log(resolved.originalLocale); // zh-CN
 * ```
 */
export function resolveDateFormatOptions(
  options: UseDateFormatOptions = {},
  contextDefaults?: DateFormatContextValue | null,
): ResolvedDateFormatOptions {
  const { locale, intlOptions, originalLocale } = mergeOptions(contextDefaults, options);
  return buildResolvedOptions(locale, intlOptions, originalLocale);
}

/**
 * @name createDateFormatter
 * @description 创建可复用的日期格式化器，适用于非 React 环境或自定义封装。
 *
 * @since 0.0.1
 *
 * @param options - 基础配置，可通过 `contextDefaults` 传入 Provider 默认值
 * @returns 一个包含 `format` 与 `resolveOptions` 方法的日期格式化器
 *
 * @example
 * ```ts
 * import { createDateFormatter } from '@internationalized/date-format';
 *
 * const formatter = createDateFormatter({
 *   dateStyle: 'medium',
 *   timeStyle: 'short',
 *   contextDefaults: { locale: 'zh-CN', timeZone: 'Asia/Shanghai' },
 * });
 *
 * const result = formatter.format(new Date('2024-05-20T00:00:00Z'));
 * console.log(result.formattedValue); // 2024/5/20 上午8:00
 * console.log(result.resolvedOptions.locale); // zh-CN
 * ```
 */
export function createDateFormatter(
  options: CreateDateFormatOptions = {},
): DateFormatter {
  const { contextDefaults = null, ...baseOptions } = options;

  return {
    format(value: DateInput, overrideOptions?: UseDateFormatOptions): UseDateFormatResult {
      const { locale, intlOptions, originalLocale } = mergeOptions(
        contextDefaults,
        baseOptions,
        overrideOptions,
      );
      return formatValue(value, locale, intlOptions, originalLocale);
    },
    resolveOptions(overrideOptions?: UseDateFormatOptions): ResolvedDateFormatOptions {
      const { locale, intlOptions, originalLocale } = mergeOptions(
        contextDefaults,
        baseOptions,
        overrideOptions,
      );
      return buildResolvedOptions(locale, intlOptions, originalLocale);
    },
  };
}
