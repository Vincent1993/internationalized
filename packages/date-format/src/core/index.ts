export { DateFormatProvider, useDateFormatContext } from './context';
export { createDateFormatter, resolveDateFormatOptions } from './formatter';
export type {
  CreateDateFormatOptions,
  DateFormatContextValue,
  DateFormatter,
  DateFormatProviderProps,
  ResolvedDateFormatOptions,
  UseDateFormatOptions,
  UseDateFormatResult,
  DateInput,
} from './types';
