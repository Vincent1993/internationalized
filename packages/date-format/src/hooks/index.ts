export { useDateFormat } from './useDateFormat';
export type { UseDateFormatOptions, DateFormatter, UseDateFormatResult } from '../core/types';
