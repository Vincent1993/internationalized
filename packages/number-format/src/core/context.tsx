import { createContext, useContext, useMemo } from 'react';
import type { NumberFormatContextValue, NumberFormatProviderProps } from './types';
import { mergeRules } from './utils/rules';

const NumberFormatContext = createContext<NumberFormatContextValue | null>(null);

/**
 * @name NumberFormatProvider
 * @description 为应用提供全局的数字格式化配置上下文，支持所有 `Intl.NumberFormat`
 * 选项与指标规则合并。
 *
 * @since 0.0.1
 *
 * @example
 * ```tsx
 * import { NumberFormatProvider, NumberFormat, AutoMetricNumber } from '@your-package/number-format';
 *
 * // 推荐用法：使用 options 对象
 * function App() {
 *   return (
 *     <NumberFormatProvider
 *       options={{
 *         locale: "zh-CN",
 *         useGrouping: true,
 *         minimumFractionDigits: 2
 *       }}
 *     >
 *       <Dashboard />
 *     </NumberFormatProvider>
 *   );
 * }
 *
 * // 带指标规则的配置
 * function AppWithRules() {
 *   const options = {
 *     locale: "zh-CN",
 *     useGrouping: true,
 *     rules: [
 *       // 精确匹配规则
 *       { name: 'revenue', options: { style: 'currency', currency: 'CNY' } },
 *       { name: 'profit', options: { style: 'currency', currency: 'CNY' } },
 *       { name: 'cost', options: { style: 'currency', currency: 'CNY' } },
 *
 *       // 正则匹配规则
 *       { pattern: '.*_rate$', options: { style: 'percent', maximumFractionDigits: 2 } },
 *       { pattern: '.*_ratio$', options: { style: 'percent', maximumFractionDigits: 1 } },
 *       { pattern: '.*_count$', options: { style: 'decimal', useGrouping: true } },
 *       { pattern: '.*_permille$', options: { style: 'per-mille' } },
 *       { pattern: '.*_permyriad$', options: { style: 'per-myriad' } },
 *       { pattern: '.*_pp$', options: { style: 'percentage-point' } },
 *       { pattern: '.*_cn_upper$', options: { style: 'cn-upper' } },
 *     ]
 *   };
 *
 *   return (
 *     <NumberFormatProvider options={options}>
 *       <MetricDashboard />
 *     </NumberFormatProvider>
 *   );
 * }
 *
 * // 嵌套 Provider（局部覆盖全局配置）
 * function NestedProviders() {
 *   return (
 *     <NumberFormatProvider options={{ locale: "zh-CN", useGrouping: true }}>
 *       <div>
 *         <h1>Global Config</h1>
 *         <NumberFormat value={1234.56} />
 *
 *         <NumberFormatProvider options={{ locale: "en-US", style: "currency", currency: "USD" }}>
 *           <div>
 *             <h2>Local Override</h2>
 *             <NumberFormat value={1234.56} />
 *           </div>
 *         </NumberFormatProvider>
 *       </div>
 *     </NumberFormatProvider>
 *   );
 * }
 * ```
 */
export const NumberFormatProvider = ({ children, options }: NumberFormatProviderProps) => {
  const parentValue = useNumberFormatContext();

  const value = useMemo<NumberFormatContextValue | null>(() => {
    if (!options) {
      return parentValue;
    }

    if (!parentValue) {
      return options;
    }

    const mergedValue: NumberFormatContextValue = {
      ...parentValue,
      ...options,
      rules: mergeRules(parentValue.rules || [], options.rules || []),
    };

    if (mergedValue.rules?.length === 0) {
      delete mergedValue.rules;
    }

    return mergedValue;
  }, [parentValue, options]);

  return <NumberFormatContext.Provider value={value}>{children}</NumberFormatContext.Provider>;
};

/**
 * @name useNumberFormatContext
 * @description React Hook，用于获取当前的数字格式化上下文配置，主要供内部 Hook
 * 与组件使用，也可用于自定义封装。
 *
 * @since 0.0.1
 *
 * @returns 当前上下文配置；若不存在 Provider 则返回 `null`
 */
export const useNumberFormatContext = (): NumberFormatContextValue | null => {
  return useContext(NumberFormatContext);
};
