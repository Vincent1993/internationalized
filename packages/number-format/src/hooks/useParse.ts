/**
 * @file 数字解析 Hook
 * @description 提供数字解析功能的 React Hook
 */

import { useMemo } from 'react';
import { NumberParser, type ParseOptions, type ParseResult } from '../core/parser';
import { useNumberFormatContext } from '../core/context';
import type { UseFormatOptions } from '../core/types';

/**
 * 解析 Hook 选项
 */
export interface UseParseOptions extends ParseOptions {
  /** 继承的格式化选项 */
  formatOptions?: UseFormatOptions;
}

/**
 * 解析 Hook 返回结果
 */
export interface UseParseResult {
  /** 解析函数 */
  parse: (input: string) => ParseResult;
  /** 解析器实例 */
  parser: NumberParser;
  /** 当前解析选项 */
  options: Required<ParseOptions>;
}

/**
 * @name useParse
 * @description 提供数字解析能力的 React Hook，会复用上下文的 locale、style 等配置。
 *
 * @since 0.0.1
 *
 * @param options - 解析选项
 * @returns 包含 `parse` 函数与解析器实例的对象
 *
 * @example
 * ```tsx
 * function InputForm() {
 *   const { parse } = useParse({
 *     style: 'currency',
 *     currency: 'USD'
 *   });
 *
 *   const [value, setValue] = useState('');
 *
 *   const handleChange = (input: string) => {
 *     const result = parse(input);
 *     if (result.success) {
 *       setValue(result.value.toString());
 *     }
 *   };
 *
 *   return <input onChange={(e) => handleChange(e.target.value)} />;
 * }
 * ```
 */
export function useParse(options: UseParseOptions = {}): UseParseResult {
  const context = useNumberFormatContext();

  // 合并上下文选项和传入选项
  const mergedOptions = useMemo(() => {
    const contextOptions = context || {};
    const formatOptions = options.formatOptions || {};

    return {
      locale: options.locale || formatOptions.locale || contextOptions.locale,
      style: options.style || formatOptions.style || contextOptions.style,
      currency: options.currency || formatOptions.currency || contextOptions.currency,
      notation: options.notation || formatOptions.notation || contextOptions.notation,
      strict: options.strict,
      groupSeparator: options.groupSeparator,
      decimalSeparator: options.decimalSeparator,
    };
  }, [options, context]);

  // 创建解析器实例
  const parser = useMemo(() => {
    return new NumberParser(mergedOptions);
  }, [mergedOptions]);

  // 解析函数
  const parse = useMemo(() => {
    return (input: string): ParseResult => parser.parse(input);
  }, [parser]);

  // 返回完整的解析选项
  const finalOptions = useMemo((): Required<ParseOptions> => {
    // 从解析器中获取实际使用的选项
    return parser.getOptions();
  }, [parser]);

  return {
    parse,
    parser,
    options: finalOptions,
  };
}
