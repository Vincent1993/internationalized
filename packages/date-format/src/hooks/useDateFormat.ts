import { useMemo } from 'react';
import { createDateFormatter } from '../core/formatter';
import { useDateFormatContext } from '../core/context';
import type { DateFormatter, UseDateFormatOptions } from '../core/types';

/**
 * @name useDateFormat
 * @description React Hook，用于创建日期时间格式化器。Hook 会合并 `DateFormatProvider`
 * 提供的上下文选项，并基于 `Intl.DateTimeFormat` 返回可复用的格式化与选项解析函数。
 *
 * @since 0.0.1
 *
 * @param options - 覆盖默认行为的格式化选项，支持所有 `Intl.DateTimeFormatOptions`
 * @returns 一个包含 `format` 与 `resolveOptions` 的日期格式化器对象
 *
 * @example
 * ```tsx
 * import { useDateFormat } from '@internationalized/date-format';
 *
 * function Clock() {
 *   const formatter = useDateFormat({
 *     timeStyle: 'medium',
 *     dateStyle: 'short',
 *     timeZone: 'Asia/Shanghai',
 *   });
 *
 *   const now = new Date();
 *   const result = formatter.format(now);
 *
 *   return <time dateTime={result.date?.toISOString()}>{result.formattedValue}</time>;
 * }
 * ```
 */
export function useDateFormat(options: UseDateFormatOptions = {}): DateFormatter {
  const contextDefaults = useDateFormatContext();

  return useMemo(() => {
    return createDateFormatter({ ...options, contextDefaults });
  }, [options, contextDefaults]);
}
