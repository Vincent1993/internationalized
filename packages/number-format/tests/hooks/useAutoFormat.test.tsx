import { renderHook } from '@testing-library/react';
import React from 'react';
import { useAutoFormat } from '../../src/hooks/useAutoFormat';
import { NumberFormatProvider } from '../../src/core/context';
import type { MetricFormatRule } from '../../src/core/types';

describe('hooks/useAutoFormat', () => {
  const mockRules: MetricFormatRule[] = [
    {
      pattern: /percentage|percent|rate/i,
      options: { style: 'percent', maximumFractionDigits: 2 },
    },
    {
      pattern: /price|cost|amount|money/i,
      options: { style: 'currency', currency: 'CNY' },
    },
    {
      pattern: /count|number|total/i,
      options: { style: 'decimal', maximumFractionDigits: 0 },
    },
  ];

  it('应该返回NumberFormatter对象', () => {
    const { result } = renderHook(() => useAutoFormat({ name: 'test_metric', rules: mockRules }));

    expect(result.current).toHaveProperty('format');
    expect(result.current).toHaveProperty('resolveOptions');
    expect(typeof result.current.format).toBe('function');
    expect(typeof result.current.resolveOptions).toBe('function');
  });

  describe('规则匹配测试', () => {
    it('应该匹配percentage规则', () => {
      const { result } = renderHook(() =>
        useAutoFormat({ name: 'conversion_percentage', rules: mockRules }),
      );

      const formatResult = result.current.format(0.1234);
      const resolved = result.current.resolveOptions();

      expect(resolved.style).toBe('percent');
      expect(resolved.originalStyle).toBe('percent');
      expect(formatResult.formattedValue).toBe('12.34%');
    });

    it('应该匹配currency规则', () => {
      const { result } = renderHook(() =>
        useAutoFormat({ name: 'product_price', rules: mockRules }),
      );

      const formatResult = result.current.format(199.99);
      const resolved = result.current.resolveOptions();

      expect(resolved.style).toBe('currency');
      expect(resolved.currency).toBe('CNY');
      expect(formatResult).toMatchSnapshot();
    });

    it('应该匹配count规则', () => {
      const { result } = renderHook(() => useAutoFormat({ name: 'user_count', rules: mockRules }));

      const formatResult = result.current.format(1234.56);
      const resolved = result.current.resolveOptions();

      expect(resolved.style).toBe('decimal');
      expect(resolved.maximumFractionDigits).toBe(0);
      expect(formatResult.formattedValue).toBe('1,235'); // 四舍五入到整数
    });

    it('应该在无匹配规则时使用默认配置', () => {
      const { result } = renderHook(() =>
        useAutoFormat({
          name: 'unknown_metric',
          rules: mockRules,
        }),
      );

      const formatResult = result.current.format(123.456789);
      const resolved = result.current.resolveOptions();

      expect(resolved.style).toBe('decimal');
      expect(formatResult.formattedValue).toBe('123.457');
    });
  });

  describe('与Context集成测试', () => {
    const contextRules: MetricFormatRule[] = [
      {
        pattern: /temperature/i,
        options: { style: 'unit', unit: 'celsius' },
      },
    ];

    const wrapper = ({ children }: { children: React.ReactNode }) => (
      <NumberFormatProvider
        options={{
          locale: 'en-US',
          rules: contextRules,
          useGrouping: false,
        }}
      >
        {children}
      </NumberFormatProvider>
    );

    it('应该合并Context和局部rules', () => {
      const localRules: MetricFormatRule[] = [
        {
          pattern: /speed|velocity/i,
          options: { style: 'unit', unit: 'kilometer-per-hour' },
        },
      ];

      const { result } = renderHook(
        () =>
          useAutoFormat({
            name: 'current_temperature',
            rules: localRules,
          }),
        { wrapper },
      );

      const resolved = result.current.resolveOptions();

      // 应该匹配Context中的temperature规则
      expect(resolved.style).toBe('unit');
      expect(resolved.unit).toBe('celsius');
    });

    it('应该局部rules优先于Context rules', () => {
      const localRules: MetricFormatRule[] = [
        {
          pattern: /temperature/i,
          options: { style: 'unit', unit: 'fahrenheit' }, // 覆盖Context规则
        },
      ];

      const { result } = renderHook(
        () =>
          useAutoFormat({
            name: 'room_temperature',
            rules: localRules,
          }),
        { wrapper },
      );

      const resolved = result.current.resolveOptions();

      // 局部规则应该优先
      expect(resolved.unit).toBe('fahrenheit');
    });
  });

  describe('边缘情况测试', () => {
    it('应该处理空规则数组', () => {
      const { result } = renderHook(() =>
        useAutoFormat({
          name: 'test_metric',
          rules: [],
        }),
      );

      const formatResult = result.current.format(123.45);
      expect(formatResult.formattedValue).toBe('123.45');
    });

    it('应该处理未定义的name', () => {
      const { result } = renderHook(() =>
        useAutoFormat({
          name: '',
          rules: mockRules,
        }),
      );

      const resolved = result.current.resolveOptions();
      expect(resolved.style).toBe('decimal');
    });

    it('应该处理复杂的pattern匹配', () => {
      const complexRules: MetricFormatRule[] = [
        {
          pattern: /^(user|customer)_(count|number)$/i,
          options: { style: 'decimal', useGrouping: true },
        },
      ];

      const { result: matchResult } = renderHook(() =>
        useAutoFormat({ name: 'user_count', rules: complexRules }),
      );

      const { result: noMatchResult } = renderHook(() =>
        useAutoFormat({ name: 'user_count_total', rules: complexRules }),
      );

      expect(matchResult.current.resolveOptions().useGrouping).toBeTruthy();
      expect(noMatchResult.current.resolveOptions().useGrouping).toBe('always');
    });

    it('应该处理无效的规则配置', () => {
      const invalidRules = [
        {
          pattern: /test/i,
          options: { style: 'currency' }, // 缺少required currency
        },
      ] as MetricFormatRule[];

      const { result } = renderHook(() =>
        useAutoFormat({ name: 'test_metric', rules: invalidRules }),
      );

      // 应该fallback并不抛出错误
      expect(() => result.current.format(123)).not.toThrow();
    });
  });

  describe('规则优先级测试', () => {
    it('应该使用最后一个匹配的规则', () => {
      const overlappingRules: MetricFormatRule[] = [
        {
          pattern: /price/i,
          options: { style: 'currency', currency: 'USD' },
        },
        {
          pattern: /price/i,
          options: { style: 'currency', currency: 'EUR' },
        },
      ];

      const { result } = renderHook(() =>
        useAutoFormat({ name: 'product_price', rules: overlappingRules }),
      );

      const resolved = result.current.resolveOptions();
      expect(resolved.currency).toBe('EUR'); // 最后一个规则生效
    });

    it('应该支持规则选项与局部选项合并', () => {
      const rules: MetricFormatRule[] = [
        {
          name: 'merge_test',
          options: { style: 'decimal', maximumFractionDigits: 2 },
        },
      ];

      const { result } = renderHook(() =>
        useAutoFormat({
          name: 'merge_test',
          rules,
        }),
      );

      const resolved = result.current.resolveOptions();
      expect(resolved.style).toBe('decimal'); // 来自规则
      expect(resolved.maximumFractionDigits).toBe(2); // 来自规则
    });
  });

  describe('快照测试', () => {
    it('应该为不同指标类型生成正确格式', () => {
      const testCases = [
        { name: 'conversion_rate', value: 0.1234 },
        { name: 'product_price', value: 299.99 },
        { name: 'user_count', value: 1234.56 },
        { name: 'unknown_metric', value: 42.195 },
      ];

      testCases.forEach(({ name, value }) => {
        const { result } = renderHook(() => useAutoFormat({ name, rules: mockRules }));

        const formatResult = result.current.format(value);
        expect(formatResult).toMatchSnapshot(`auto-format-${name}-${value}`);
      });
    });
  });
});
