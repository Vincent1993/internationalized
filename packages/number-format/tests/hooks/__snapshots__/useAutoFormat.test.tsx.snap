// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`hooks/useAutoFormat > 快照测试 > 应该为不同指标类型生成正确格式 > auto-format-conversion_rate-0.1234 1`] = `
{
  "formattedValue": "12.34%",
  "isNaN": false,
  "parts": [
    {
      "type": "integer",
      "value": "12",
    },
    {
      "type": "decimal",
      "value": ".",
    },
    {
      "type": "fraction",
      "value": "34",
    },
    {
      "type": "percentSign",
      "value": "%",
    },
  ],
  "resolvedOptions": {
    "includeSign": false,
    "locale": "zh-CN",
    "maximumFractionDigits": 2,
    "minimumFractionDigits": 0,
    "minimumIntegerDigits": 1,
    "notation": "standard",
    "numberingSystem": "latn",
    "originalStyle": "percent",
    "roundingIncrement": 1,
    "roundingMode": "halfExpand",
    "roundingPriority": "auto",
    "signDisplay": "auto",
    "style": "percent",
    "trailingZeroDisplay": "auto",
    "useGrouping": "always",
  },
  "sign": {
    "isNegative": false,
    "isPositive": true,
    "isZero": false,
    "numeric": 1,
  },
  "value": 0.1234,
}
`;

exports[`hooks/useAutoFormat > 快照测试 > 应该为不同指标类型生成正确格式 > auto-format-product_price-299.99 1`] = `
{
  "formattedValue": "¥299.99",
  "isNaN": false,
  "parts": [
    {
      "type": "currency",
      "value": "¥",
    },
    {
      "type": "integer",
      "value": "299",
    },
    {
      "type": "decimal",
      "value": ".",
    },
    {
      "type": "fraction",
      "value": "99",
    },
  ],
  "resolvedOptions": {
    "currency": "CNY",
    "currencyDisplay": "symbol",
    "currencySign": "standard",
    "includeSign": false,
    "locale": "zh-CN",
    "maximumFractionDigits": 2,
    "minimumFractionDigits": 2,
    "minimumIntegerDigits": 1,
    "notation": "standard",
    "numberingSystem": "latn",
    "originalStyle": "currency",
    "roundingIncrement": 1,
    "roundingMode": "halfExpand",
    "roundingPriority": "auto",
    "signDisplay": "auto",
    "style": "currency",
    "trailingZeroDisplay": "auto",
    "useGrouping": "always",
  },
  "sign": {
    "isNegative": false,
    "isPositive": true,
    "isZero": false,
    "numeric": 1,
  },
  "value": 299.99,
}
`;

exports[`hooks/useAutoFormat > 快照测试 > 应该为不同指标类型生成正确格式 > auto-format-unknown_metric-42.195 1`] = `
{
  "formattedValue": "42.195",
  "isNaN": false,
  "parts": [
    {
      "type": "integer",
      "value": "42",
    },
    {
      "type": "decimal",
      "value": ".",
    },
    {
      "type": "fraction",
      "value": "195",
    },
  ],
  "resolvedOptions": {
    "includeSign": false,
    "locale": "zh-CN",
    "maximumFractionDigits": 3,
    "minimumFractionDigits": 0,
    "minimumIntegerDigits": 1,
    "notation": "standard",
    "numberingSystem": "latn",
    "originalStyle": "decimal",
    "roundingIncrement": 1,
    "roundingMode": "halfExpand",
    "roundingPriority": "auto",
    "signDisplay": "auto",
    "style": "decimal",
    "trailingZeroDisplay": "auto",
    "useGrouping": "always",
  },
  "sign": {
    "isNegative": false,
    "isPositive": true,
    "isZero": false,
    "numeric": 1,
  },
  "value": 42.195,
}
`;

exports[`hooks/useAutoFormat > 快照测试 > 应该为不同指标类型生成正确格式 > auto-format-user_count-1234.56 1`] = `
{
  "formattedValue": "1,235",
  "isNaN": false,
  "parts": [
    {
      "type": "integer",
      "value": "1",
    },
    {
      "type": "group",
      "value": ",",
    },
    {
      "type": "integer",
      "value": "235",
    },
  ],
  "resolvedOptions": {
    "includeSign": false,
    "locale": "zh-CN",
    "maximumFractionDigits": 0,
    "minimumFractionDigits": 0,
    "minimumIntegerDigits": 1,
    "notation": "standard",
    "numberingSystem": "latn",
    "originalStyle": "decimal",
    "roundingIncrement": 1,
    "roundingMode": "halfExpand",
    "roundingPriority": "auto",
    "signDisplay": "auto",
    "style": "decimal",
    "trailingZeroDisplay": "auto",
    "useGrouping": "always",
  },
  "sign": {
    "isNegative": false,
    "isPositive": true,
    "isZero": false,
    "numeric": 1,
  },
  "value": 1234.56,
}
`;

exports[`hooks/useAutoFormat > 规则匹配测试 > 应该匹配currency规则 1`] = `
{
  "formattedValue": "¥199.99",
  "isNaN": false,
  "parts": [
    {
      "type": "currency",
      "value": "¥",
    },
    {
      "type": "integer",
      "value": "199",
    },
    {
      "type": "decimal",
      "value": ".",
    },
    {
      "type": "fraction",
      "value": "99",
    },
  ],
  "resolvedOptions": {
    "currency": "CNY",
    "currencyDisplay": "symbol",
    "currencySign": "standard",
    "includeSign": false,
    "locale": "zh-CN",
    "maximumFractionDigits": 2,
    "minimumFractionDigits": 2,
    "minimumIntegerDigits": 1,
    "notation": "standard",
    "numberingSystem": "latn",
    "originalStyle": "currency",
    "roundingIncrement": 1,
    "roundingMode": "halfExpand",
    "roundingPriority": "auto",
    "signDisplay": "auto",
    "style": "currency",
    "trailingZeroDisplay": "auto",
    "useGrouping": "always",
  },
  "sign": {
    "isNegative": false,
    "isPositive": true,
    "isZero": false,
    "numeric": 1,
  },
  "value": 199.99,
}
`;
